lms/static/scripts/frontend_apps/components/BookSelector.js

```diff
@@ -0,0 +1,156 @@
+import {
+  IconButton,
+  SvgIcon,
+  TextInput,
+  TextInputWithButton,
+  Thumbnail,
+} from '@hypothesis/frontend-shared';
+import { createElement } from 'preact';
+import { useRef, useState } from 'preact/hooks';
+
+import { useService, VitalSourceService } from '../services';
+
+/**
+ * @typedef {import('../api-types').Book} Book
+ */
+
+/**
+ * @typedef BookSelectorProps
+ * @prop {Book|null} selectedBook
+ * @prop {(b: Book|null) => void} onSelectBook - Callback invoked when user selects a book
+ */
+
+/**
+ * Component that prompts a user to enter/paste a URL to a VitalSource book.
+ *
+ * @param {BookSelectorProps} props
+ */
+export default function BookSelector({ selectedBook, onSelectBook }) {
+  const vsService = useService(VitalSourceService);
+
+  const inputRef = useRef(/** @type {HTMLInputElement|null} */ (null));
+  const [isLoading, setIsLoading] = useState(false);
+
+  const [error, setError] = useState(/** @type {string|null} */ (null));
+  const previousURL = useRef(/** @type {string|null} */ (null));
+
+  /**
+   * A naive regex matcher for a VBID in a URL-like string
+   *
+   * @param {string} url
+   * @returns {string|false}
+   */
+  const bookIDFromURL = url => {
+    const bookIdPattern = /book\/([0-9A-Z-]+)\/?/;
+    const matches = url.match(bookIdPattern);
+    return !!matches && matches[1];
+  };
+
+  /**
+   * Attempt to retrieve a book by bookID (vbid) via service
+   *
+   * @param {string} bookID
+   */
+  const fetchBook = bookID => {
+    setIsLoading(true);
+    vsService
+      .fetchBook(bookID)
+      .then(onSelectBook)
+      .catch(err => setError(err.message))
+      .finally(() => setIsLoading(false));
+  };
+
+  // Respond to changes in the input field
+  const onChangeURL = () => {
+    const url = inputRef.current.value;
+    if (url && url === previousURL.current) {
+      // Do nothing if there is no real change. This situation can arise if the
+      // user "submits" a URL by typing "Enter" and then interacts elsewhere in
+      // the UI (a spurious change event can be triggered).
+      // This prevents reloading the same book.
+      return;
+    }
+    previousURL.current = url;
+
+    // Reset selected book, if any
+    onSelectBook(null);
+    setError(null);
+
+    if (url) {
+      // Don't try to fetch, or set any errors, if the input field is empty
+      const bookID = bookIDFromURL(url);
+      if (!bookID) {
+        setError("That doesn't look like a VitalSource book URL");
+        return;
+      }
+      fetchBook(bookID);
+    }
+  };
+
+  // Capture "Enter" and make sure it does not submit the whole shebang
```

Comment:
By "whole shebang" do you mean the parent `<form>` rendered by the `FilePickerApp` component?

Reply:
Yep. I will improve this comment.